apps/champions/lib/champions/battle/simulator.ex

```diff
@@ -63,187 +64,459 @@ defmodule Champions.Battle.Simulator do
+    {%{skill | effects_trigger: skill.effects_trigger - 1}, current_state}
+  end
 
-    Enum.reduce(skill.effects, current_state, fn effect, new_state ->
-      # If skill doesn't have targeting strategy, we fall back to the effects'
-      target_ids =
-        if is_nil(target_ids),
-          do: choose_targets(unit, effect, initial_step_state),
-          else: target_ids
+  # Calculate the new state of the battle after a step passes for a pending effect.
+  # If the effect is ready to be processed, we apply it.
+  defp process_step_for_effect(%{delay: 0} = effect, current_state) do
+    Logger.info("#{format_unit_name(effect.caster)}'s effect is ready to be processed")
 
-      targets_after_effect = Enum.map(target_ids, fn id -> apply_effect(effect, unit, new_state[id]) end)
+    targets_after_effect =
+      Enum.map(effect.targets, fn id ->
+        maybe_apply_effect(effect, current_state.units[id], effect.caster, current_state.step_number)
+      end)
 
-      Enum.reduce(targets_after_effect, new_state, fn target_unit, acc_state ->
-        Map.put(acc_state, target_unit.id, target_unit)
+    current_state =
+      Enum.reduce(targets_after_effect, current_state, fn target, acc_state ->
+        put_in(acc_state, [:units, target.id], target)
       end)
-    end)
+
+    Map.put(current_state, :pending_effects, List.delete(current_state.pending_effects, effect))
+  end
+
+  # Calculate the new state of the battle after a step passes for a pending effect.
+  # If the effect isn't ready to be processed, we reduce its remaining delay.
+  defp process_step_for_effect(effect, current_state) do
+    Map.put(current_state, :pending_effects, [
+      %{effect | delay: effect.delay - 1} | List.delete(current_state.pending_effects, effect)
+    ])
   end
 
-  defp choose_targets(_unit, %{targeting_strategy: nil}, _state), do: nil
+  # Check if the unit can attack this turn.
+  # For now, attacking capability is only affected by whether the unit is currently casting a skill.
+  # Later on, things like stuns will be handled here.
+  defp can_attack(unit, initial_step_state) do
+    # Check the unit is not casting anything right now
+    not Enum.any?(initial_step_state.skills_being_cast, &(&1.caster_id == unit.id))
+  end
+
+  # Check if the unit can cast their ultimate skill this step.
+  defp can_cast_ultimate_skill(unit), do: unit.energy >= @ultimate_energy_cost
+
+  # Check if the unit can cast their basic skill this step.
+  defp can_cast_basic_skill(unit), do: unit.basic_skill.remaining_cooldown <= 0
+
+  # Called when a skill being cast reaches effect_trigger 0.
+  # "Queues" the effect to be processed when its delay reaches 0.
+  defp trigger_skill_effects(skill, current_state, initial_step_state) do
+    caster = current_state.units[skill.caster_id]
 
+    # We store the caster's state in the effect in case the unit dies before the effect's delay ends
```

Comment:
We want to apply the effect if the caster dies before the delay ends? if that's not true why adding the caster to the skill if we're checking that the caster is alive in the `process_step_for_skill` method?

Reply:
1. We **do** want to apply the effect if the caster dies before the **effect delay** ends, if the effect has already been triggered. When we don't want to do that is when the effect hasn't been triggered and the unit dies, and that's handled on skill cooldown reduce.
2. I'll explain another reason for this implementation with an example:
- You receive a buff that duplicates your damage
- You cast a fireball
- The fireball takes a while to travel the screen
- Meanwhile, you lose your buff
- Your fireball finally hits its target
- The damage it deals isn't doubled because your buff expired, even though the fireball was cast when you indeed had it.
This is not the behaviour we want! I'll add another comment to mention this in the comments since it's missing

Reply:
ooooh okey, this aim to make the effects not dynamic, what i mean with this is that the modifiers will only be applied when the ability is casts and won't affect the ability later on. Sounds good!